### PR TITLE
fix: handle missing Windows npm cache trees (#3092)

### LINKED ISSUE #3092: [Bug]: Windows managed runtime cache repair 500s on a missing _npx tree, blocking the --prefer-online startup retry
### Summary

On Windows the managed-npm-runtime startup recovery can never succeed. When `npx --prefer-offline` fails to resolve the pinned agent runtime (stale cached packument → `ETARGET`), the recovery is supposed to repair the npm cache and relaunch with `--prefer-online`. The repair step instead always errors when the `_npx` execution tree for that spec is absent — which is exactly the state the recovery exists to handle — so every task launch dies with:

```text
failed to initialize ACP: agent_runtime_error: managed runtime cache repair failed with status 500
```

Root cause: `internal/agent/managedruntime/cache_windows.go` detects "child not found" with `errors.Is(err, windows.ERROR_FILE_NOT_FOUND / ERROR_PATH_NOT_FOUND)`, but `NtCreateFile` reports it as a `windows.NTStatus`, which never compares equal to a `syscall.Errno`.

### Affected area

Agent lifecycle

### Kandev version

5c5b3c6d9

### Install mode

Source build

### Environment

- OS: Windows 11 Pro 10.0.26200 (native, no WSL)
- Architecture: amd64
- Browser: Edge 151
- Node: v24.18.0
- npm: 12.0.2
- Docker/Kubernetes: not used (local executor)
- Build: native Windows build via MSYS2 UCRT64, `-tags fts5`

### Kandev configuration

- Executor type/profile: `local`, `use_worktree=false`
- Agent profile: Claude Code (`claude-acp`, managed npm runtime)
- Managed runtime pin: `@agentclientprotocol/claude-agent-acp@0.70.0` (`managed_npm_runtime_versions.json`)
- Repository source: local path (redacted)
- Relevant `KANDEV_*` env vars: none set beyond defaults

### Task/session context

Not task-specific — reproduces for every task on the `claude-acp` agent profile, on any workflow/step, across fresh starts and resumes. Observed on four consecutive launch attempts of the same task (two `resume`, one `fresh_start`), all failing identically. Task/session/execution IDs are local-only UUIDs and carry no upstream meaning, so they are omitted.

### Steps to reproduce

1. Windows host whose npm cache holds a packument for the pinned managed runtime that predates the pinned version — e.g. cached `dist-tags.latest = 0.63.0` while `managed_npm_runtime_versions.json` pins `0.70.0`. In that state the `_npx/<key>` execution tree under the npm cache does not exist for the pinned spec (`<key>` = first 16 hex chars of `sha512(<exact package spec>)`, here `fca12915ff656968`).
2. Start any task on the `claude-acp` agent profile with a `local` executor.
3. The npx child exits during npm resolution; ACP `initialize` reports `peer disconnected before response`.
4. The backend logs `retrying managed npm runtime startup with online metadata` and then fails the startup with `managed runtime cache repair failed with status 500`. The `--prefer-online` relaunch never happens.
5. Retrying from the UI (resume or fresh start) reproduces steps 3–4 verbatim.

Verifying the trigger independently, outside kandev:

```text
$ npx --yes --prefer-offline @agentclientprotocol/claude-agent-acp@0.70.0 --version
npm error code ETARGET
npm error notarget No matching version found for @agentclientprotocol/claude-agent-acp@0.70.0.

$ npx --yes --prefer-online @agentclientprotocol/claude-agent-acp@0.70.0 --version
0.70.0

$ npx --yes --prefer-offline @agentclientprotocol/claude-agent-acp@0.70.0 --version
0.70.0
```

The last line is the host-side workaround: warm the runtime online once and the pinned command resolves, after which tasks start normally. Note `npm view <spec>` is **not** sufficient — the install path uses the `application/vnd.npm.install-v1+json` cache entry, which stays stale.

### Reproducibility

Always

### Clean-state check

Not tried

(The state that matters here is the host npm cache, not the Kandev data directory — the failure survived backend restarts, new sessions, resume and fresh-start launches.)

### Expected behavior

An `ETARGET` from a stale offline packument is a recoverable condition: the recovery relaunches with `--prefer-online`, npm revalidates metadata, the pinned version installs and the session initializes. A cache-repair step that finds nothing to remove is a no-op, not a failure — the execution tree is legitimately absent whenever the version was never installed, and what is stale is the cached packument, not the tree. Repair failure should also not be the thing that suppresses the relaunch that actually fixes the problem.

### Actual behavior

`stopAndRepairManagedRuntime` is a hard gate: any repair error returns `needsFailure=true` and the startup is published as failed (`managed_runtime_startup.go:272-282`), so the `--prefer-online` relaunch (`managed_runtime_recovery.go:38`) is never attempted.

The repair fails inside `internal/agent/managedruntime/cache_windows.go:removeNpxExecutionTree`:

```go
targetHandle, err := openManagedRuntimeDirectoryRelative(npxHandle, key)
if errors.Is(err, windows.ERROR_PATH_NOT_FOUND) || errors.Is(err, windows.ERROR_FILE_NOT_FOUND) {
    return nil
}
```

`NtCreateFile` reports a missing child as `windows.NTStatus`, which has no `Unwrap`/`Is`, so neither `errors.Is` can ever match. Verified against `golang.org/x/sys v0.45.0` with a standalone probe — a temp dir containing `_npx`, opening an absent child through the same `openManagedRuntimeHandle` call:

```text
missing-child err        = Object Name not found.
type                     = windows.NTStatus
errors.Is FILE_NOT_FOUND = false
errors.Is PATH_NOT_FOUND = false
NTSTATUS                 = 0xC0000034   (STATUS_OBJECT_NAME_NOT_FOUND)
```

So the function returns `open npm execution tree: Object Name not found.`, and the agentctl handler maps it to HTTP 500 (`internal/agentctl/server/api/managed_runtime.go:37-44`).

The same NTSTATUS quirk is already known one level up: there is an `os.Lstat` pre-check for the `_npx` directory with a comment describing exactly this behavior, but no equivalent guard for `_npx/<key>`.

Net effect is a catch-22 — repair only succeeds when the execution tree already exists, yet the tree cannot exist when the version was never installed, which is the only situation where the `ETARGET` recovery runs. On Windows the recovery path is dead code for its primary case.

### Logs and artifacts

Backend, one launch attempt (all four are identical):

```text
agent process started   command="npx --yes --prefer-offline @agentclientprotocol/claude-agent-acp@0.70.0"
initializing ACP session   agent_type=claude-acp   native_session_resume=true
sending ACP initialize request
[agentctl-launcher stderr] INFO connection closed component=acp-conn cause="peer connection closed"
ACP initialize failed   agent_type=claude-acp
  error="initialize failed: ACP initialize handshake failed:
         {"code":-32603,"message":"Internal error",
          "data":{"error":"peer disconnected before response"}}"
session initialization failed
retrying managed npm runtime startup with online metadata   startup_generation=2
failed to start agent process
  error="failed to initialize ACP: agent_runtime_error: managed runtime cache repair failed with status 500"
handling agent stopped
failed to recover session   action=resume
  error="session not ready after resume: session failed: failed to initialize ACP:
         agent_runtime_error: managed runtime cache repair failed with status 500"
```

Browser: `Task failed to start — failed to initialize ACP: agent_runtime_error: managed runtime cache repair failed with status 500`.

The stale cache entry that triggers it, read out of `_cacache`:

```text
key:  make-fetch-happen:request-cache:https://registry.npmjs.org/@agentclientprotocol%2fclaude-agent-acp
resHeaders.date:  Tue, 28 Jul 2026 12:37:31 GMT
resHeaders.cache-control:  public, max-age=300
cached dist-tags:  {'latest': '0.63.0'}
cached versions:   55, newest 0.63.0
cached has 0.70.0: False
```

No log line explains the 500 itself: `handleManagedRuntimeCacheRepair` logs a bare `"managed runtime cache repair failed"` with no error field, and the HTTP body is a fixed string. The npx stderr that drove the ETARGET classification is not logged either, so diagnosing this required reading the source rather than the logs.

### Last known good version

unknown

The failure surfaces once the managed-runtime pin moves ahead of the host's cached metadata — here with the `0.70.0` pin from 358a41a8a. Any host whose cached packument predates a pin bump is affected, so it is not tied to one kandev version.

### Extra context

Suggested fixes:

1. Treat an NTSTATUS "not found" as absence in `cache_windows.go` — either an `os.Lstat` pre-check on `_npx/<key>` mirroring the existing `_npx` guard, or map through `NTStatus.Errno()` before `errors.Is`. Worth auditing the other `errors.Is(err, windows.ERROR_*)` checks around the `NtCreateFile` calls in that file (root and `_npx` opens, `openManagedRuntimeEntryRelative`) for the same mismatch.
2. Make the repair best-effort in `stopAndRepairManagedRuntime`: "nothing to remove" — and arguably any repair failure — should log and continue to the `--prefer-online` relaunch rather than fail the startup.
3. Log the underlying error in `handleManagedRuntimeCacheRepair`, and the npx stderr behind the ETARGET classification.
4. Add a regression test for "`_npx` exists, `<key>` absent". The current `TestRemoveNpxExecutionTreeIsIdempotentAndRejectsUnsafeTargets` only covers a cache root with no `_npx` directory at all, so the early `Lstat` short-circuits before reaching the failing branch.

Optional hardening: revalidating pinned managed-runtime metadata — or using `--prefer-online` for the first launch of a newly pinned version — would avoid entering this recovery path after every version bump.

Possibly related: #2259 (`Agent launch depends on npx resolution with no timeout or fallback`, closed) covers npx resolution fragility at launch, but not the repair-gate failure described here.

### Before submitting

- [x] I can reproduce this using the steps above.
- [x] I removed secrets, tokens, private repository URLs, private prompts, and customer data.
- [x] I searched existing issues and did not find a duplicate.

### PR DESCRIPTION
**Pull request by carlosflorencio**
> [!TIP]
> **PR walkthrough:** [Open the visual walkthrough](https://walkthrough.kandev.ai/pr/3094/1765b2c11985.html)

Windows managed npm cache repair treated a missing `_npx/<key>` execution tree as an error because `NtCreateFile` returned `NTStatus` values that did not match Win32 errors. The repair now normalizes those statuses and retries cleanly while preserving exact-tree and path-safety behavior.

## Validation

- `go test ./internal/agent/managedruntime` passed with 36 tests.
- `GOOS=windows GOARCH=amd64 CGO_ENABLED=0 go test -c` compiled the Windows test binary.
- `make -n test-windows` confirmed the focused native regression command.
- `python3 scripts/lint-spec-files.py --all` passed.
- Native Windows execution is configured in CI because this workspace runs Linux.

## Possible Improvements

Low risk. Native Windows execution remains CI-only in this environment.

## Checklist

- [ ] I have performed a self-review of my code.
- [ ] I have manually tested my changes and they work as expected.
- [ ] My changes have tests that cover the new functionality and edge cases.
- [ ] If my change touches UI files (`apps/web/`), I have added or updated Playwright e2e tests in `apps/web/e2e/` and verified them with `make test-e2e`.
- [ ] I checked whether this affects public docs in `docs/public/**` and updated them or noted why no docs change is needed.

Closes #3092

<a href="https://cubic.dev/pr/kdlbs/kandev/pull/3094?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: apps/backend/Makefile [hunk 1]
```diff
@@ -351,12 +351,14 @@ test:
...
 ## .github/workflows/backend-tests.yml. The broader suite uses Unix-only
 ## shell commands in test fixtures and would need GOOS guards to be portable;
-## this narrow scope is the regression coverage for the conpty/job-object
-## bug class. Tests that genuinely cannot run on Windows skip themselves via
-## runtime.GOOS guards in source.
+## this narrow scope is the regression coverage for Windows-specific behavior.
+## Tests that genuinely cannot run on Windows skip themselves via runtime.GOOS
+## guards in source.
 test-windows:
 	@echo "Running Windows-sensitive packages..."
```
[the remark this answers is not in the document]
Comment (carlosflorencio) [reply apps/backend/Makefile:358]: Addressed in commit `1765b2c119`: the Make target now explains that the focused filter avoids symlink-dependent tests on Windows and directs future Windows-only tests to this command.
